SDWebImage/SDWebImageDownloaderOperation.m

```diff
@@ -329,93 +319,37 @@ - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)data
-                // oriented incorrectly sometimes. (Unlike the image born of initWithData
-                // in didCompleteWithError.) So save it here and pass it on later.
-#if SD_UIKIT || SD_WATCH
-                _orientation = [[self class] orientationFromPropertyValue:(orientationValue == -1 ? 1 : orientationValue)];
-#endif
-            }
-        }
-        
-        if (_width + _height > 0 && !finished) {
-            // Create the image
-            CGImageRef partialImageRef = CGImageSourceCreateImageAtIndex(_imageSource, 0, NULL);
-            
-#if SD_UIKIT || SD_WATCH
-            // Workaround for iOS anamorphic image
-            if (partialImageRef) {
-                const size_t partialHeight = CGImageGetHeight(partialImageRef);
-                CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
-                CGContextRef bmContext = CGBitmapContextCreate(NULL, _width, _height, 8, _width * 4, colorSpace, kCGBitmapByteOrderDefault | kCGImageAlphaPremultipliedFirst);
-                CGColorSpaceRelease(colorSpace);
-                if (bmContext) {
-                    CGContextDrawImage(bmContext, (CGRect){.origin.x = 0.0f, .origin.y = 0.0f, .size.width = _width, .size.height = partialHeight}, partialImageRef);
-                    CGImageRelease(partialImageRef);
-                    partialImageRef = CGBitmapContextCreateImage(bmContext);
-                    CGContextRelease(bmContext);
-                }
-                else {
-                    CGImageRelease(partialImageRef);
-                    partialImageRef = nil;
-                }
-            }
-#endif
-            
-            if (partialImageRef) {
-#if SD_UIKIT || SD_WATCH
-                UIImage *image = [UIImage imageWithCGImage:partialImageRef scale:1 orientation:_orientation];
-#elif SD_MAC
-                UIImage *image = [[UIImage alloc] initWithCGImage:partialImageRef size:NSZeroSize];
-#endif
-                CGImageRelease(partialImageRef);
-                NSString *key = [[SDWebImageManager sharedManager] cacheKeyForURL:self.request.URL];
-                UIImage *scaledImage = [self scaledImageForKey:key image:image];
-                if (self.shouldDecompressImages) {
-                    image = [UIImage decodedImageWithImage:scaledImage];
-                }
-                else {
-                    image = scaledImage;
+        if (!self.progressiveCoder) {
+            // We need to create a new instance for progressive decoding to avoid conflicts
+            id<SDWebImageProgressiveCoder> progressiveCoder = (id<SDWebImageProgressiveCoder>)[[SDWebImageCodersManager sharedInstance] coderWithCondition:^BOOL(id<SDWebImageCoder>  _Nonnull coder) {
+                if ([coder conformsToProtocol:@protocol(SDWebImageProgressiveCoder)]) {
+                    if ([((id<SDWebImageProgressiveCoder>)coder) canIncrementalDecodeData:imageData]) {
+                        return YES;
+                    }
```

Comment:
@dreampiggy I don't really like the API `coderWithCondition`. Since `SDWebImageCodersManager` exposes all the coders via an array, we don't need that new API.
I suggest that this "reuse coder logic" is limited to this case only. I like storing the progressive coder inside the downloader operation.
But we can do that by simply looping through the list of coders:
```
if (!self.progressiveCoder) {
  // We need to create a new instance for progressive decoding to avoid conflicts
  for (id<SDWebImageCoder> in [SDWebImageCodersManager sharedInstance].coders) {
    if ([coder conformsToProtocol:@protocol(SDWebImageProgressiveCoder)] && 
        [((id<SDWebImageProgressiveCoder>)coder) canIncrementalDecodeData:imageData]) {
      self.progressiveCoder = [[[progressiveCoder class] alloc] init];
    }
  }
}
```

Reply:
Previously I think that the `for-in loop` in `coders` may be not always obey the priority order.  So if this order gurantee is never change (coders Array first object contains the highest priority and last object contains the lowest one), we can remove that help method.

And also, I advice to point this out clearly in that property doc comments. So that user would not do reversed enumeration.

Reply:
I'm not too worried about this order thing - it's something the user will config just once, so any potential problems will be visible. I also would let him empty the coders array, it's the expected behaviour, not do any magic behind scenes.
Also, ideally only one coder from the array can decode a certain type of data, so we should be good.